drivers/spi/spi_pico_pio.c

```diff
@@ -0,0 +1,403 @@
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_pico_pio);
+
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "spi_context.h"
+
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+
+#include <hardware/pio.h>
+#include "hardware/clocks.h"
+
+#define PIO_CYCLES     (4)
+#define PIO_FIFO_DEPTH (4)
+
+struct spi_pico_pio_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *pin_cfg;
+	struct gpio_dt_spec clk_gpio;
+	struct gpio_dt_spec mosi_gpio;
+	struct gpio_dt_spec miso_gpio;
+	const uint32_t clock_freq;
+};
+
+struct spi_pico_pio_data {
+	struct spi_context spi_ctx;
+	uint32_t tx_count;
+	uint32_t rx_count;
+	PIO pio;
+	size_t pio_sm;
+	int dfs;
+};
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_0_cpha_0, 0, 1,
+			    /*     .wrap_target */
+			    0x6101, /*  0: out    pins, 1         side 0 [1] */
+			    0x5101, /*  1: in     pins, 1         side 1 [1] */
+				/*     .wrap */
+);
+
+RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_1_cpha_1, 0, 2,
+			    /*     .wrap_target */
+			    0x7021, /*  0: out    x, 1            side 1 */
+			    0xa101, /*  1: mov    pins, x         side 0 [1] */
+			    0x5001, /*  2: in     pins, 1         side 1 */
+				/*     .wrap */
+);
+
+static float spi_pico_pio_clock_divisor(const struct spi_pico_pio_config *dev_cfg,
+	uint32_t spi_frequency)
+{
+	return (float)dev_cfg->clock_freq / (float)(PIO_CYCLES * spi_frequency);
+}
+
+static uint32_t spi_pico_pio_maximum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / PIO_CYCLES;
+}
+
+static uint32_t spi_pico_pio_minimum_clock_frequency(const struct spi_pico_pio_config *dev_cfg)
+{
+	return dev_cfg->clock_freq / (PIO_CYCLES * UINT16_MAX);
+}
+
+static inline bool spi_pico_pio_transfer_ongoing(struct spi_pico_pio_data *data)
+{
+	return spi_context_tx_on(&data->spi_ctx) || spi_context_rx_on(&data->spi_ctx);
+}
+
+static inline void spi_pico_pio_sm_put8(PIO pio, uint sm, uint8_t data)
+{
+	/* Do 8 bit accesses on FIFO, so that write data is byte-replicated. This */
+	/* gets us the left-justification for free (for MSB-first shift-out) */
+	io_rw_8 *txfifo = (io_rw_8 *)&pio->txf[sm];
+
+	*txfifo = data;
+}
+
+static inline uint8_t spi_pico_pio_sm_get8(PIO pio, uint sm)
+{
+	/* Do 8 bit accesses on FIFO, so that write data is byte-replicated. This */
+	/* gets us the left-justification for free (for MSB-first shift-out) */
+	io_rw_8 *rxfifo = (io_rw_8 *)&pio->rxf[sm];
+
+	return *rxfifo;
+}
+
+static int spi_pico_pio_configure(const struct spi_pico_pio_config *dev_cfg,
+				  struct spi_pico_pio_data *data, const struct spi_config *spi_cfg)
+{
+	const struct gpio_dt_spec *miso = NULL;
+	const struct gpio_dt_spec *mosi = NULL;
+	const struct gpio_dt_spec *clk = NULL;
+	pio_sm_config sm_config;
+	uint32_t offset;
+	uint32_t wrap_target;
+	uint32_t wrap;
+	int cpol;
```

Comment:
It should better to uint32. (Zephyr don't use much `int` excluding return code')